### DISCUSSION TITLE
Mismatched type issue

### DISCUSSION THREAD

**Discussion by deephog** [PINTO0309/PINTO_model_zoo#266] [Q&A]
Hi, do you know what may cause this kind of mismatched type [issue](https://github.com/NVIDIA/TensorRT/issues/2131)?

**Comment by PINTO0309** [comment]
I have never done INT8 quantization of ONNX before running trtexec.

**Reply by PINTO0309** [reply]
https://github.com/PINTO0309/simple-onnx-processing-tools/releases/tag/1.0.41

**Reply by PINTO0309** [reply]
https://github.com/PINTO0309/simple-onnx-processing-tools

**Comment by deephog** [comment]
Two further questions:

1. What do you recommend as a more practical way to quantize a model and convert it to tensorrt, if I am working on a Stereo-Depth model that usually involve tricky layers such as "gridSample", "Where", "Gather" and etc..  

2. And also, to this specific issue, forget about the quantization part, I checked the onnx file of the model after simplification, the layer that reports a mismatched type has an input node of INT64 and a output node of INT64, how come itself is expecting a flaot32? And if it is caused by either ONNX generation of Pytorch or ONNX-simplifier, and I wanted to manually revise the properties using your tool, what should I revise,  If I couldn't even find a "float32" to work on?

Thanks!

**Reply by PINTO0309** [reply]
>What do you recommend as a more practical way to quantize a model and convert it to tensorrt, if I am working on a Stereo-Depth model that usually involve tricky layers such as "gridSample", "Where", "Gather" and etc..

I think the only way to do this is to replace all of them with OPs that support TensorRT. Before converting to ONNX.

>And also, to this specific issue, forget about the quantization part, I checked the onnx file of the model after simplification, the layer that reports a mismatched type has an input node of INT64 and a output node of INT64, how come itself is expecting a flaot32? And if it is caused by either ONNX generation of Pytorch or ONNX-simplifier, and I wanted to manually revise the properties using your tool, what should I revise, If I couldn't even find a "float32" to work on?

I don't know anything about it because I don't own the ONNX file before quantization.

**Comment by PINTO0309** [comment]
https://github.com/megvii-research/Sparsebit

**Reply by deephog** [reply]
Thank you for the suggestion, I will try their own tool to see if it works better.

Actually now quantization is no longer the immediate issue, I found that the model works very differently even under FP16. It is reasonable since a disparity estimation model will definitely be more sensitive to minor weight value changes. 

Do you know any good method of doing FP16 aware training? Pytorch seems only to provide mixed precision training which is of no use to this situation.

**Reply by PINTO0309** [reply]
I am not aware of any way to train regular means using methods other than mixing precision that avoids the effects of batch normalization, etc.
https://pytorch.org/docs/stable/notes/amp_examples.html
